### PR TITLE
Improve external gravity

### PR DESCRIPTION
**Pull request by hyschive**
This PR improves the implementation of external acceleration and potential. Major improvements are listed below.

- Replace the options (`OPT__GRAVITY_TYPE` and `OPT__EXTERNAL_POT`) by (`OPT__SELF_GRAVITY`, `OPT__EXT_ACC`, and `OPT__EXT_POT`) (see [here](https://github.com/gamer-project/gamer/wiki/Gravity#opt__self_gravity--0off-1on--1)).
- Support external potential in the model `HYDRO`.
- The gravitational potential in the code now stores the **total** potential instead of **self-gravity** potential unless otherwise specified.
- Simplify the external gravity interface (see [here](https://github.com/gamer-project/gamer/wiki/Gravity#external-accelerationpotential)). Now one only needs to set a single function pointer in a problem initialization function.
- Support external potential in the test problem `Plummer`.
- Add a user-specified function before the Poisson solver (see `src/SelfGravity/Poi_UserWorkBeforePoisson.cpp`).

Unfinished tasks:
- [ ] Set external potential from an input table.
- [ ] Port to wave dark matter simulations.

@hfhsieh Can you help review it and, especially, see whether it fits your GREP implementation?

### CODE REVIEW

File: src/Init/Init_GAMER.cpp [hunk 1]
```diff
@@ -244,7 +243,7 @@ void Init_GAMER( int *argc, char ***argv )
       } // for (int lv=0; lv<NLEVEL; lv++)
 
       if ( MPI_Rank == 0 )    Aux_Message( stdout, "%s ... done\n", "Calculating gravitational potential" );
-   } // if ( OPT__GRAVITY_TYPE == GRAVITY_SELF  ||  OPT__GRAVITY_TYPE == GRAVITY_BOTH )
+   } // if ( OPT__SELF_GRAVITY_TYPE ||  OPT__EXT_POT )
```
Comment (hfhsieh) [inline src/Init/Init_GAMER.cpp:246]: missing one space before ||

File: src/TestProblem/Hydro/Plummer/ExtAcc_Plummer.cpp [hunk 2]
```diff
@@ -8,36 +8,81 @@
...
+extern double Plummer_R0;
+extern double Plummer_Center[3];
+extern double Plummer_ExtAccMFrac;
+
+//-------------------------------------------------------------------------------------------------------
+// Function    :  SetExtAccAuxArray_Plummer
+// Description :  Set the auxiliary array ExtAcc_AuxArray[] used by ExtAcc_Plummer()
+//
+// Note        :  1. Invoked by Init_ExtAcc_Plummer()
+//                2. AuxArray[] has the size of EXT_ACC_NAUX_MAX defined in Macro.h (default = 10)
```
Comment (hfhsieh) [inline src/TestProblem/Hydro/Plummer/ExtAcc_Plummer.cpp:27]: default = 20 for `EXT_POT_NAUX_MAX`, `EXT_ACC_NAUX_MAX`, and `EOS_NAUX_MAX`. same typos in other files.

File: src/Init/Init_GAMER.cpp (hunk 1 above)
Comment (hyschive) [reply src/Init/Init_GAMER.cpp]: Fixed. Thanks!

File: src/TestProblem/Hydro/Plummer/ExtAcc_Plummer.cpp (hunk 2 above)
Comment (hyschive) [reply src/TestProblem/Hydro/Plummer/ExtAcc_Plummer.cpp]: 1. Fixed.
2. It's always defined in `CUPOT.h`. `Init_TestProb_ELBDM_ExtPot.cpp` doesn't really use `EXT_POT_NAUX_MAX`. Does compilation fail in some of your tests?

Comment (hyschive) [reply src/TestProblem/Hydro/Plummer/ExtAcc_Plummer.cpp]: 2. Fixed

**Comment by hfhsieh** [comment]
I think it ready for merge.

**Review by hfhsieh — APPROVED** [review]

**Pull request merged**